Added support for extracting images from Webpack assets stats

dimchez opened the pull request:
This PR adds support for extracting images processed by Webpack, related to #18 
- Added `images` array to resulting JSON which contains `name` and `path` for images processed by Webpack
- Wrapping extracted chunks with static assets into `chunks` object to separate entry points related data from images array

Since generated JSON structure changes I suggest bumping major version if this PR gets merged into master.

dimchez committed:
Updated generated structure to include chunk assets under 'chunk' key in resulting JSON
Fixed expectations in unit tests
Added 'tdd' task in npm scripts

dimchez committed:
Added images array to JSON output
Added tests for handling images and updated existing tests
Added file-loader as dev dependency for tests
Updated README.md according to changes in JSON output format
Updated CHANGELOG.md with information about changed/added functionality

kossnocorp commented:
@dimchez great stuff!

kossnocorp commented on readme.md:
```diff
@@ -30,16 +30,27 @@ The output is a JSON object in the form:
 
-  * `"bundle_name"` is the name of the bundle (the key of the entry object in your webpack config, or "main" if your entry is an array).
-  * `"asset_kind"` is the camel-cased file extension of the asset
+  * `chunks` is an object that contains all entry points listed in your webpack config
```
@dimchez if you don't mind I'd prefer to name it `entries` because docs says:

>  "contains all **entry points** listed in your webpack config"

dimchez replied:
@kossnocorp Sure, I'll update PR to use `entries` everywhere instead of `chunks`.

kossnocorp replied:
@dimchez thanks!

kossnocorp commented on readme.md:
```diff
@@ -30,16 +30,27 @@ The output is a JSON object in the form:
+        "name": "hello.svg",
+        "path": "/public/path/to/asset/hello.svg"
+      }
+    ]
```
@dimchez I want to make it a bit more generic, so it would be possible attach any kind of assets. It's reasonable to rename `images` to `assets`. WDYT?

kossnocorp replied:
If you wish, you can finalize it (we need to add an ability to override `/\.(jpe?g|png|gif|svg)$/`). Otherwise, I'll merge PR and update the code by myself. Thank you for your help!

dimchez replied:
@kossnocorp my idea was to separate assets by type. I am using extracted `images` array to resolve URLs to images when rendering on server-side. I guess having more generic `assets` would work as well. Just less data to iterate over when looking up a URL.

dimchez replied:
@kossnocorp I've added `imagesRegex` option for overriding default regex.

kossnocorp replied:
> Just less data to iterate over when looking up a URL.

I see your point, but since we have `/\.(jpe?g|png|gif|svg)$/)` as default pattern, you'll have the same amount of data.

Let's name it:
- `assets` for the key.
- `assetsRegex` for the option.

dimchez replied:
@kossnocorp Alright, I've updated PR.

dimchez committed:
replaced chunks with entries

dimchez committed:
Added possibility to override imagesRegex
Added tests for overriding of imagesRegex option
Fixed typo in CommonChunksPlugin name

dimchez committed:
updated readme with info about imagesRegex option

dimchez committed:
Replaced images with assets, imagesRegex option with assetsRegex option
Updated tests
Updated readme and changelog accordingly

kossnocorp commented:
Perfect!

kossnocorp commented:
@sporto could you add me to npm package owners? My username there is `kossnocorp`.

sporto commented:
@kossnocorp     should be done now

kossnocorp commented:
@sporto thank you a lot! 

@dimchez give me few days to test everything and I'll publish the new version.

lennerd commented:
Whats the progress here?

lennerd commented:
I tested the current master and I'm not quite sure, how this is supposed to work. From the current index.js:

``` js
output.assets = stats.assets.filter(function (asset) {
  return self.options.assetsRegex.test(asset.name)
}).map(function (asset) {
  return { name: asset.name, path: assetPath + asset.name }
});
```

What happens, if we use something like `[hash]`  inside our file loader name parameter? The asset `name` and `path` would be the same containing the hash except for the suffixed asset path.

This way it is very hard to parse the JSON manifest file on the other side without extra configuration for parsing the file name and excluding the hash for example.

Sample output from my setup:

``` json
{
  "entries": {
    "main": {
      "js": "main-[sha512:hash:base64:7].js",
      "css": "main-[sha512:hash:base64:7].css"
    }
  },
  "assets": [
    {
      "name": "/images/3qSDcel.jpg",
      "path": "/images/3qSDcel.jpg"
    }
  ]
}
```
